fossas/fossa-cli
# Support Ruby GIT remote dependencies

# Overview

We have not been handling Git remote dependencies in Ruby projects. We have been stripped the remote and making them `GemType` deps. This PR handles this by making them `GitType` deps and also handling an edge case where some dependencies in the `DEPENDENCIES` section have a `!` suffix if they have a remote other than the main one listed for RubyGems.

## Acceptance criteria

The main criteria was testing on the impacted user's project and seeing that the dependencies were correctly detected. This has been solved

## Testing plan
There were two things I needed to test and I went about it the same way
1. `GIT` remote deps are detected properly
2. Dependencies in `DEPENDENCIES` with `!` suffix have it stripped **and** that this does not negatively impact detecting other dependencies.

Testing this was relatively simple
1. Ensure that the impacted user's project scanned successfully
2. Search for other popular and random other projects on GitHub through "site:github.com inurl:Gemfile.lock" and ensure that they scan properly.

## Risks

Before this change, it was impossible to detect Git remote deps, so there should be no risk there except for user's not expecting GIT dependencies to show up.
The riskiest part is handling the `!` suffix convention. The convention has documentation publicly and is used by several repositories (the linked ANE ticket and the one linked in the code comment). The main risk here is that there are other impacts of handling the "!" suffix that I am not aware of. As of now, I am fully convinced this is the correct move, based on reviewing ~50 GitHub repositories and observing no conflicts.

## References

- [ANE-1624](https://fossa.atlassian.net/browse/ANE-1624): Issue where GIT remote deps in Ruby projects aren't showing up

## Checklist



[ANE-1624]: https://fossa.atlassian.net/browse/ANE-1624?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
